### ISSUE TITLE
Use aligned data for SIMD code wherever possible 

### ISSUE THREAD

**Issue by ihhub** [ihhub/penguinV#255]
We have SIMD code in `src/image_function_simd.cpp` file. It works fine. We want to have the best performance as possible. There are multiple places where we could improve speed. We use `_mm_storeu_si128` and `_mm256_storeu_si256` to copy results from simd data types into normal (`uint32_t`) types. Normal types are unaligned. We could align them and use the same operations but for unaligned versions (`_mm_store_si128` and `_mm256_store_si256`). Such operations are faster than original.

**Comment by ihhub** [comment]
@0x72D0 you might have an interest in this ;) There are 3 or 4 places only to cover.

**Comment by 0x72D0** [comment]
oh this project sounds fun! But do you have an idea how can we do that? We have to make a different alignment size for SSE (16-bit) and AVX (32-bit). Also, we need to align the vector itself so we can't  simply use memalign

**Comment by 0x72D0** [comment]
I think the easiest way to go is to make or use a specific allocator to std::vector

**Comment by ihhub** [comment]
My first assumption is regarding to the code for loading final data into variable. We cannot assume that std::vector is aligned correctly but we could makes checks in the function to see an aligment of input and output data before running the real code. The biggest problem is that instricts are not functions so you cannot use function pointers.

**Comment by ihhub** [comment]
You could see that every function uses unaligned data load and save which is a big drawback in performance.

**Comment by 0x72D0** [comment]
we cannot assume that std::vector is aligned, but according to the template prototype: `template < class T, class Alloc = allocator<T> > class vector;` we can pass an allocator to the vector, so if we align the data with the allocator, we know that the data is align

**Comment by 0x72D0** [comment]
some people on stack overflow have tried this: https://stackoverflow.com/questions/12942548/making-stdvector-allocate-aligned-memory

**Comment by ihhub** [comment]
I mean that end user (somebody else) would not care about this thing while calling the function.

**Comment by 0x72D0** [comment]
yeah that's true...

**Comment by 0x72D0** [comment]
maybe we can give to the end user a "special" kind of vector, by doing a `using image_vector = vector<T, allocator_simd>` and we change the function call to projectionProfile and Accumulate

**Comment by ihhub** [comment]
No-no. I'm talking about general solution. We use std::vector only for one function while we have many of them :) We should concentrate our focus of image data load and save.

**Comment by 0x72D0** [comment]
but the image also use an std::vector, no?

**Comment by ihhub** [comment]
No, it's a raw allocated array. Additionally, we handle ROI (region of interest) which could be smaller than original image size. Even if you allocate an image with correct alignment it could be a situation when ROI start position is not aligned.

**Comment by 0x72D0** [comment]
oh I understand now... ROI make everything more difficult... the only way I can think is copying the ROI in an aligned memory space... but I think, that will hurt the performance...

**Comment by ihhub** [comment]
Yes, it is. Another idea which comes into my mind is that we can create a template function with 2 typenames. One typename parameter is load function for input data, another one is save function for output data. We verify the alignment at the very beginning and call needed function. You could try this and measure performance after.

**Comment by 0x72D0** [comment]
I'm not really sure about what you're talking about...
```
template <typename a, typename b>
void foo {
    (*a)() // load function
    (*b)() // save function
}
```
 something like this? But in a template we can't put function as a typename, so I'm not sure to understand...

**Comment by ihhub** [comment]
Let's assume we create a template function:
```cpp
template <typename _alignment>
simd loadSSE(simd * data);
```
Then we could have a specification of this template as:
```cpp
template<>
simd loadSSE<aligned>(simd * data)
{
    return _mm256_load_si256( data );
}

template<>
simd loadSSE<unaligned>(simd * data)
{
    return _mm256_loadu_si256( data );
}
```

Then you could make a real function which present in the code, let's say `AbsoluteDifference` to be a template:
```cpp
template <typename  _inputAlignment, typename  _outputAlignment>
void AbsoluteDifference(...)
...
loadSSE<_inputAlignment>( src1 );
```

But in this situation you must make your initial template functions for simd load/save as inline.

**Comment by 0x72D0** [comment]
@ihhub with the new memory allocator, do we still need to check if the memory is aligned or we can consider it's always aligned?

**Comment by ihhub** [comment]
Yes, we need to check. Memory allocator is an optional feature.

**Comment by ihhub** [comment]
I think the alternative solution is to add a check in memory allocator whether current array is allocated there. Then we could know if all images are allocated within it which makes all data aligned as needed.
